fxtrace.1
'\" t
.\"     Title: FXTRACE
.\"    Author: Kisel Jan
.\" Generator: vim
.\"      Date: 25/04/2015
.\"    Manual: Release 0.3
.\"    Source: Release 0.3
.\"  Language: English
.\"
.TH "FXTRACE" "1" "13/05/2015" "Release 0.3" "Release 0.3"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
fxtrace \- tool for tracing file acceess
.SH "SYNOPSIS"
.HP \w'\fBfxtrace\fR\ 'u
\fBfxtrace\fR [\fB\-l\fR|\fB\-\-log\fR logfile] [\fB\-m\fR|\fB\-\-mode\fR access] [\fB-p\fR|\fB\-\-prefix\fR prefix] [\fB\-d\fR|\fB\-\-debug] [\-\-] cmd [...]
.SH "DESCRIPTION"
.PP
\fBFxtrace\fR
is utility for tracing file access. It uses technique of replacing library calls with LD_PRELOAD and may be seen as alternative to strace.
.SH "OPTIONS"
.PP
\fB\-l\fR|\fB\-\-log\fR \fIlogfile\fR
.RS 4
Log file access to given logfile. By default, fxtrace writes traces to fxtrace.log in current directory.
.RE
.RS 4
Events being logged right after their performing. Each line of output file has following format:
.RS 2

access file

.RE .RE
.RS 4
Where access is one or several chars of rwxs, which stand for read, write, execute and stat correspondingly.\n
.BL
.RE
.RS 4
Do not confuse this with chmod(1): these rwxs about \fIhow\fR file is being used, not which attributes it has\&.
.RE
.PP
\fB\-m|\-\-mode access\fR
.RS 4
Log only selected events:
.PP
.RS 2
- r: file opened for read or shared library loaded
.RE
.RS 2
- w: file opened for write
.RE
.RS 2
- x: file executed or shared library loaded
.RE
.RS 2
- s: stat(2) performed on file. Typical example is \`ls -l\'\&.
.RE
.PP
Any combination of following modes can be used: r, rw, wrx, etc\&.

.RE
.PP
\fB\-p\fR|\fB\-\-prefix\fR prefix\fR
.RS 4
Log only files which fullpath starts with given prefix\&.
.RE
.PP
\fB-d\fR|\fB\-\-debug\fR
.RS 4
Print all events to stderr\&.
.RE
.PP
\fB-h\fR|\fB\-\-help\fR
.RS 4
Print usage and short notes about options\&.
.RE
.SH "EXAMPLES"
.PP
Simplest usage:
.RE
.RS 2

$ fxtrace cat /etc/fstab 

.RE
.RS 2
this will log, at least /bin/cat and /etc/fstab in default logfile fxtrace.txt
.RE
.PP
More advanced usage is to provide restict-options:
.RE 
.RS 2

$ fxtrace --log /tmp/fxtrace.txt --mode wr --prefix $HOME mc

.RE
.RS 2
In this case you will see all files Midnight Commander read or write within your home directory.
.SH "AUTHOR"
.PP
Kisel Jan <jdkrot@gmail\&.com>
.SH "REPORTING BUGS"
.PP
Report fxtrace bugs to: https://github.com/dkrotx/fxtrace/issues
.RE
Fxtrace home page:      https://github.com/dkrotx/fxtrace
.SH "SEE ALSO"
.PP
strace(1), ltrace(1)


NEWS
See https://github.com/dkrotx/fotrack for comments


src/fxtrace.sh
#!/bin/bash

set -e
set -o pipefail

FXTRACE_LIB_PATH=${FXTRACE_LIB_PATH=@FXTRACE_SO_DIR@/libfxtrace.so}
FXTRACE_LOG=./fxtrace.log

usage() {
    if [[ -n $1 ]]; then
        echo "$0: wrong usage: " "$@"
    fi

    local progname=$( basename $0 )

    cat - >&2 <<ENDUSAGE
Usage: $progname [-l logfile] [-m mode] [-p prefix] [-d] cmd [...]

Options:
  -l|--log:    trace operations to specified file instead of $FXTRACE_LOG
  -m|--mode:   trace only specified operations. They are:
                 r - open for read
                 w - open for write
                 x - execute
                 s - stat(2)
 
  -p|--prefix:  log only files which fullpath starts with given prefix
  -d|--debug:   print each event to stderr

EXAMPLES:

- Simple case:
  $ $progname cat /etc/fstab # $FXTRACE_LOG will contain log

- More advanced case:
  $ $progname --log /tmp/fxtrace.txt --mode wsr --prefix \$HOME mc

ENDUSAGE
    exit ${USAGE_EXIT_CODE:-64}
}

warn() {
    echo -n "$0: " >&2
    echo -e "warning: $@" >&2
}

err() {
    echo "error: $@" >&2
    exit 1
}

extract_longopt_value() {
    local var=$1
    local value=${2#*=}

    [[ -n $value ]] || usage
    eval $var="$value"
}

###############################################################################
## MAIN
###############################################################################


while [[ $# -gt 0 ]]; do
    case "$1" in
        -l|--log)     FXTRACE_LOG=$2;     shift 2 ;;
        -m|--mode)    FXTRACE_MODE=$2;    shift 2 ;;
        -p|--prefix)  FXTRACE_PREFIX=$2;  shift 2 ;;
        -d|--debug)   FXTRACE_DEBUG=1;    shift ;;
        -h|--help*)   USAGE_EXIT_CODE=0 usage; break ;;

        
        --log=*)    extract_longopt_value FXTRACE_LOG "$1";    shift ;;
        --mode=*)   extract_longopt_value FXTRACE_MODE "$1";   shift ;;
        --prefix=*) extract_longopt_value FXTRACE_PREFIX "$1"; shift ;;

        # support GNU longops for booleans too (--debug=1). why not?
        --debug=*)   extract_longopt_value FXTRACE_DEBUG "$1";   shift ;;

        --) shift; break ;; # end of options
        -*) usage; break ;; # unknown option
        * ) break ;;        # hit 'cmd'
    esac
done


# check library only when it's path given
if [[ $( expr index "$FXTRACE_LIB_PATH" / || true ) -ne 0 ]]; then
    if [[ ${FXTRACE_LIB_PATH:0:1} = '/' ]]; then
        readlink -e "$FXTRACE_LIB_PATH" >/dev/null 2>&1 || err "$FXTRACE_LIB_PATH nor found or not proper link"
    else
        # do not allow relative path since cmd may do `cd somethere && exec something'
        abspath=$( readlink -e "$FXTRACE_LIB_PATH" ) || err "requred lib ($FXTRACE_LIB_PATH) not found"
        FXTRACE_LIB_PATH=$abspath
    fi
fi

[[ $# -gt 0 ]] || usage "no cmd param found"
[[ -n $FXTRACE_LOG ]] || usage "--log must not be empty"

# FXTRACE_LOG must be absolute path, since subprocess will reopen logfile
[[ ${FXTRACE_LOG:0:1} = '/' ]] || FXTRACE_LOG=$( readlink -f "$FXTRACE_LOG" )

if [[ -n $FXTRACE_MODE ]]; then
    rest=$( echo -n "$FXTRACE_MODE" | tr -d rwxs )
    [[ -z $rest ]] || usage "extra symbol(s) in --mode ($rest)"
    export FXTRACE_MODE
fi

if [[ -n $FXTRACE_PREFIX ]]; then
    [[ ${FXTRACE_PREFIX:0:1} = '/' ]] || FXTRACE_PREFIX=$( readlink -m "$FXTRACE_PREFIX" )
    export FXTRACE_PREFIX
fi

[[ -n $FXTRACE_DEBUG ]] && export FXTRACE_DEBUG

# we have to clear logfile since fxtrace.so will only append to it
: >$FXTRACE_LOG
export FXTRACE_LOG

env LD_PRELOAD="$FXTRACE_LIB_PATH" "$@"
